### PR TITLE
Fix encoding & hardhat test

### PR DESCRIPTION
**Pull request by tommyrharper**
## Description

Add a hardhat test for conditional order signatures, and fix a tiny bug in the SC `ConditionalOrderHashLib.hash` function.

## Related PR
Fix/addition to [PR](https://github.com/Kwenta/smart-margin-v3/pull/15)

## Motivation and Context

Front-end was struggling to get signatures to match, reached out for help, I got pulled down the rabbit hole and decided to try and debug.

I found these two stack-overflow posts with the same issue:
- https://stackoverflow.com/questions/75019935/how-to-define-typehash-for-eip712-typed-data-signing-with-nested-struct-in-solid
- https://ethereum.stackexchange.com/questions/151513/eip712-typeddata-encoding-with-nested-array-of-structs-returning-wrong-signer

And also another example that uses the new approach:
- https://gist.github.com/sambacha/d2fb1d45b2a8cf0696c2297125006a6d

In both of those posts, the solution used `encodePacked` for the array of hashes before hashing that array instead of `encode`. Honestly I couldn't tell this myself from reading the EIP, so I wrote a hardhat test to verify, and it fails without this change, and passes with it.

[This thread](https://forum.openzeppelin.com/t/abi-encode-vs-abi-encodepacked/2948/2) questions the reasoning behind this. `bb_terk` says:

> Looking at the example in EIP-712 ([https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.js 599](https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.js)), I see that dynamic arrays/strings are not padded, instead of static types that are encoded with padding. Maybe the reason is that a dynamic array could be always different from another one, instead of a static type that could have the same rappresentation in memory. I don’t find another reason for that.

That's the closest I could come to a reason for using `encodePacked` other than the fact it works and others use it in this scenario. The [actual EIP itself](https://eips.ethereum.org/EIPS/eip-712#definition-of-typed-structured-data-%F0%9D%95%8A) says the following:

> The array values are encoded as the keccak256 hash of the concatenated encodeData of their contents (i.e. the encoding of SomeType[5] is identical to that of a struct containing five members of type SomeType).

This phraseology is pretty opaque to me, if anything it makes me think to use `encode` over `encodePacked`. Hence it is possible that Viem and ethers have incorrectly interpreted the EIP. It is either that, or I am not reading it correctly or am missing some details in it.

Note the following change was due to a `reverted with panic code 0x32 (Array accessed at an out-of-bounds or negative index)` error when testing an array of conditions with hardhat:
```solidity
bytes32[] memory hashedConditions = new bytes32[](co.conditions.length);
```

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge fmt`?
- [x] Ran `forge snapshot`?
- [x] Ran `forge test`?

### CODE REVIEW

File: test/Signature.test.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  describe("Signature checks", function () {
+    it("The engine deployed successfully", async function () {
+      const { engine } = await loadFixture(bootstrapSystem);
+
+      expect(await engine.getAddress()).to.exist;
+    });
+
+    it("Signature is verified", async () => {
```
Comment (JaredBorders) [inline test/Signature.test.ts:62]: very nicely done ☑️

File: test/utils/ConditionalOrderSignature.sol
Comment (JaredBorders) [inline test/utils/ConditionalOrderSignature.sol:1]: jw why you did not change:
```
bytes32[] memory hashedConditions;
```
to
```
bytes32[] memory hashedConditions = new bytes32[](co.conditions.length);
```

You did this in the library

**Review by JaredBorders — APPROVED** [review]
I had a single question but otherwise everything LGTM ✨

File: test/utils/ConditionalOrderSignature.sol
Comment (tommyrharper) [reply test/utils/ConditionalOrderSignature.sol:1]: I am surprised that the tests don't fail without this change. I have updated it to use `co.conditions.length)` - this leaves some confusion for me, as I believe a memory array is fixed to its initialisation length. Hence it should fail on the first iteration of the loop 🤷‍♂️

**Pull request merged**